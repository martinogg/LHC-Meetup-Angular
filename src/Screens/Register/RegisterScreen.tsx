import React, { Component } from 'react';
import { Alert, Button, Platform, StyleSheet, Text, View } from 'react-native';
import { NavigationActions, NavigationScreenProp, StackActions } from 'react-navigation'
import { Image, KeyboardAvoidingView } from 'react-native';

import { AppStyles } from '../../AppStyles'
import LoginForm from '../../Components/LoginForm/LoginForm';
import FirebaseConnection from '../../Helpers/FirebaseConnection'

interface Props {
    navigation: NavigationScreenProp<any, any>,
    screenProps: {
        firebaseConnection: FirebaseConnection
    }
}

export class RegisterScreen extends Component<Props> {
    public render() {
        //<Image resizeMode="contain" style={styles.logo} source={require('../../components/images/logo-dark-bg.png')} />
        return (
            <KeyboardAvoidingView behavior="padding" style={AppStyles.container}>
                <View style={styles.loginContainer}>

                </View>
                <View style={styles.formContainer}>
                    <LoginForm onLoginCallback={(username, password) => this.register(username, password)} actionButtonText='REGISTER' />
                </View>
                <Button title="Edit Screen"
                    onPress={() => this.goToEditScreen()}
                />
            </KeyboardAvoidingView>
        );
    }

    private register(username: string, password: string) {

        this.props.screenProps.firebaseConnection.register(username, password).then(() => {

            this.showAlert('Registration OK!');
            if (this.props.screenProps.firebaseConnection.isLoggedIn()) {

                this.goToHomeScreen()
            }
        }, (fail) => {

            this.showAlert('Registration Fail. Reason: ' + fail);
        })
    }

    private showAlert(message: string) {
        
        Alert.alert(message);
    }

    private goToHomeScreen() {

        this.props.navigation.dispatch(
            StackActions.reset({
                index: 0,
                actions: [NavigationActions.navigate({ routeName: 'Home' })]
            })
        )
    }


}

const styles = StyleSheet.create({
    loginContainer: {
        alignItems: 'center',
        flexGrow: 1,
        justifyContent: 'center'
    },
    logo: {
        position: 'absolute',
        width: 300,
        height: 100
    },
    title: {
        color: "#FFF",
        marginTop: 120,
        width: 180,
        textAlign: 'center',
        opacity: 0.9
    }
})