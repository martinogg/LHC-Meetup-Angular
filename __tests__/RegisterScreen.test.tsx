import React from 'react';
import 'react-native';

import { shallow } from 'enzyme';
import { Button, Alert } from 'react-native'

import { RegisterScreen } from '../src/Screens/Register/RegisterScreen';

import renderer from 'react-test-renderer';

const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn(),
    replace: jest.fn(),
    dispatch: jest.fn()
  },
  ...props
});

it('should display RegisterScreen with no errors', () => {

  let props: any;
  props = createTestProps({});

  const navigation = { navigate: jest.fn() };
  expect(renderer.create(<RegisterScreen {...props} />)).toMatchSnapshot();
});

test('test goToHomeScreen function', () => {

  let props: any;
  props = createTestProps({});

  const wrapper = shallow(<RegisterScreen {...props} />);
  const sut: any = wrapper.instance()
  sut.goToHomeScreen()

  expect(props.navigation.dispatch).toHaveBeenCalledTimes(1)
  expect(props.navigation.dispatch).toHaveBeenCalledWith({ "actions": [{ "routeName": "Home", "type": "Navigation/NAVIGATE" }], "index": 0, "key": null, "type": "Navigation/RESET" })
});


test('test login function on pass', async () => {

  const username = "a"
  const password = "b"

  const registerFunc = (usernamePassedIn: string, passwordPassedIn: string): Promise<string> => {

    expect(usernamePassedIn).toBe(username)
    expect(passwordPassedIn).toBe(password)

    return new Promise((resolve, reject) => {
      resolve()
    })
  }

  let props: any;
  props = createTestProps({
    screenProps: {
      firebaseConnection: {
        register: registerFunc,
        isLoggedIn: () => { return true }
      }
    }
  });

  const wrapper = shallow(<RegisterScreen {...props} />);
  const sut: any = wrapper.instance()
  let returnedAlertText = ''

  sut.showAlert = (alertText: string) => {

    returnedAlertText = alertText
  }

  sut.goToHomeScreen = jest.fn()

  await sut.register(username, password)
  expect(returnedAlertText).toBe('Registration OK!')

  expect(sut.goToHomeScreen).toHaveBeenCalledTimes(1)
});

test('test login function on fail', async () => {

  const username = "a"
  const password = "b"

  const registerFunc = (usernamePassedIn: string, passwordPassedIn: string): Promise<string> => {

    expect(usernamePassedIn).toBe(username)
    expect(passwordPassedIn).toBe(password)

    return new Promise((resolve, reject) => {
      reject('A reason')
    })
  }

  let props: any;
  props = createTestProps({
    screenProps: {
      firebaseConnection: {
        register: registerFunc
      }
    }
  });

  const wrapper = shallow(<RegisterScreen {...props} />);
  const sut: any = wrapper.instance()
  let returnedAlertText = ''

  sut.showAlert = (alertText: string) => {

    returnedAlertText = alertText
  }

  await sut.register(username, password)
  expect(returnedAlertText).toBe('Registration Fail. Reason: A reason')
});

test('test showAlert', () => {

  jest.mock('Alert', () => {
    return {
      alert: jest.fn()
    }
  });

  let props: any;
  props = createTestProps({});

  const wrapper = shallow(<RegisterScreen {...props} />);
  const sut: any = wrapper.instance()

  const showMessage = 'aMessage'
  sut.showAlert(showMessage)

  expect(Alert.alert).toBeCalledTimes(1)
  expect(Alert.alert).toBeCalledWith(showMessage)
})