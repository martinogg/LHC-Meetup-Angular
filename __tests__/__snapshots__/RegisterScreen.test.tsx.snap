// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display RegisterScreen with no errors 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#2c3e50",
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
        "justifyContent": "center",
      }
    }
  />
  <View>
    <View
      style={
        Object {
          "padding": 20,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        keyboardType="email-address"
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Email"
        placeholderTextColor="rgba(225,225,225,0.7)"
        rejectResponderTermination={true}
        returnKeyType="next"
        style={
          Object {
            "backgroundColor": "rgba(225,225,225,0.2)",
            "color": "#fff",
            "height": 40,
            "marginBottom": 10,
            "padding": 10,
          }
        }
        underlineColorAndroid="transparent"
      />
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Password"
        placeholderTextColor="rgba(225,225,225,0.7)"
        rejectResponderTermination={true}
        returnKeyType="go"
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "rgba(225,225,225,0.2)",
            "color": "#fff",
            "height": 40,
            "marginBottom": 10,
            "padding": 10,
          }
        }
        underlineColorAndroid="transparent"
      />
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#2980b6",
            "margin": 10,
            "opacity": 1,
            "paddingVertical": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontWeight": "700",
              "textAlign": "center",
            }
          }
        >
          REGISTER
        </Text>
      </View>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Edit Screen
      </Text>
    </View>
  </View>
</View>
`;
